### DISCUSSION TITLE
How to associate Terminal opened for task with CustomExecution

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#1256] [Extension Development QnA] [answered]
I'm implementing a `vscode.CustomExecution` to integrate a custom build system. I want to add some buttons to interact with a task backed by this custom implementation into the terminal toolbar. I know how to put buttons into the terminal toolbar in general and about how `when` clauses can be used in `package.json` to control when these buttons show up. My problem is I don't know how to tell which `vscode.Terminal` is the one opened for my `vscode.CustomExecution` so I can set some context keys when it is active and get back to my custom task instance from the command handlers.

Basically I want to write code like this
```ts
vscode.window.onDidChangeActiveTerminal(term => {
    void vscode.commands.executeCommand('setContext', `isMyCustomTask`, isTerminalOfMyCustomTask(term))
})

vscode.commands.registerCommand("commandForCustomTask", () => {
    const term = vscode.window.activeTerminal
    const exec = myCustomExecutionInTerminal(term)
    actOnCustomExecution(exec)
  })
```
but don't know how to do the `isTerminalOfMyCustomTask` and `myCustomExecutionInTerminal`.

I was hoping I could use `creationOptions` of the terminal
```ts
export interface Terminal {
  readonly creationOptions: Readonly<TerminalOptions | ExtensionTerminalOptions>;
}
export interface ExtensionTerminalOptions {
  pty: Pseudoterminal;
}
```
to correlate execution with the terminal because I expected the `creationOptions` to be the `ExtensionTerminalOptions` which would contain the reference to the `vscode.Pseudoterminal` I created in my custom task execution (So I could simply have a `Map<vscode.Pseudoterminal, MyCustomExecution>` somewhere that gets populated by the custom execution so it can be looked up later). This is not the case though, the the terminal is created with `TerminalOptions` (I think this because the object I see has properties `shellPath` and `shellArgs` which only exist in `vscode.TerminalOptions`) - not sure if this is a bug and should be reported as such on its own, an oversight or I'm missing something here...

**Comment by theReynald** [comment]
Hi vaclavHala,

To associate a vscode.Terminal with a vscode.CustomExecution and identify it later, you can maintain a mapping between custom executions and terminals using a unique identifier (e.g., a UUID) for each execution. 

When creating a terminal for your custom execution, include this unique identifier in the terminal's name or environment variables. 

Then, implement functions to check if an active terminal corresponds to one of your custom executions by matching this identifier in the terminal's name or environment variables. 

This approach allows you to set context keys or perform actions specific to the terminal associated with your custom execution.

I got this suggestion from copilot so for more specific examples, you can have it update your code to experiment. Let me know how it goes & reopen if you have more questions.

/r

**Reply by vaclavHala** [reply]
Hi, thanks for the suggestion but it does not help me. 

You see I don't have control over how the Terminal is created - vscode creates it not me. The first time I see the Terminal is inside the `onDidOpenTerminal` at which point I can't modify its environment nor name (in fact I don't even know if I should modify that particular Terminal, maybe it is one opened by user with real shell).

Likewise I don't have full control over the Terminal's name because when the `vscode.Task` I'm given to resolve already has name set from user whatever I put into the resolved Task I return from `resolveTask` seems to be ignored in favor of the user supplied name.

Could you maybe point me to some working code sample? From just your description I don't think what you are saying would work, or if I'm missing something don't see what. Thanks

**Reply by olguzzar** [reply]
Have you checked out this [terminal extension sample](https://github.com/microsoft/vscode-extension-samples/tree/main/terminal-sample)?

**Reply by vaclavHala** [reply]
Hello @olguzzar , yes I have and I read through it again now to be sure but I can't see how it addresses my question.

The sample shows how to create a terminal and how to interact with it once you have a reference to it. Neither of those are my question. Since I'm writing a `vscode.CustomExecution` I'm not in control of how/when the terminal gets created, vscode does that on its own. My problem is how to recognize that terminal seen inside `vscode.window.onDidChangeActiveTerminal` is the terminal created by vscode for my `vscode.CustomExecution` (see original question for exact code I'm trying to write).

If the sample addresses this in some way I don't see could you please elaborate on what you have in mind exactly? Thanks

**Reply by olguzzar** [reply]
Okay so if I understand correctly, you need to be able to figure out what the origin of the currently active terminal is - and if it the origin is from your custom task, that's when you need to show your custom buttons?

**Reply by vaclavHala** [reply]
Well to be precise my requirement is a bit simpler than determining origin for an arbitrary terminal (i.e. being able to recognize all the ways terminal can be created - by a task, by user clicking in the UI, by running a command, ...), I just need to be able to implement a predicate like this
```ts
terminalBelongsToCustomExecution(terminal: vscode.Terminal, execution: vscode.CustomExecution): boolean
```
but essentially what you described is correct yes.

**Reply by robconery** [reply]
@isidorn Is this something you can help with?

**Reply by isidorn** [reply]
I do not know. Maybe @meganrogge can help

**Reply by meganrogge** [reply]
when you create the `CustomExecution`, do you await the callback that will return the `Pseudoterminal` that task is running in? that's how you could populate the map.

**Reply by meganrogge** [reply]
https://github.com/microsoft/vscode/blob/30d42ee6c0116d7b94610938c8c95e38d0530f23/src/vscode-dts/vscode.d.ts#L8690-L8697

**Reply by vaclavHala** [reply]
@meganrogge thanks for the suggestion, but that does not solve my problem - I have the Pseudoterminal and know how to use that to control the terminal, but I don't know what `vscode.Terminal` instance is associated with the `vscode.Pseudoterminal` that I have. So this is not enough info to populate the map - I have only one half of the mapping

Let me share some more code which will hopefully make it clearer.

This is a very small extension that demonstrates my problem:
```js
function activate(context) {
    vscode.tasks.registerTaskProvider("test", {
        provideTasks() {
            return []
        },
        resolveTask(task) {
            return new vscode.Task(task.definition, task.scope, task.name, task.source, new vscode.CustomExecution(def => {
                const _onDidWrite = new vscode.EventEmitter()
                /** @type {vscode.Pseudoterminal} */
                const pty = {
                    onDidWrite: _onDidWrite.event,
                    open(){
                        _onDidWrite.fire("Hello")
                    },
                    close(){}
                }
                // here I have the Pseudoterminal and CustomExecution, but not Terminal
                return Promise.resolve(pty)
            }))
        }
    })
    vscode.window.onDidOpenTerminal(term => {
        // here I have the Terminal but not the Pseudoterminal that controls it
        console.log(term)
    })
}
```

When inside the callback where I created the PTY I don't know how to get the Terminal instance associated with the PTY, in fact the associated Terminal may not even exist yet as vscode maybe only creates it using the PTY once I return it from the callback. On the other hand once I get reference to the Terminal (inside the `onDidOpenTerminal` listener) I don't know how to get the PTY from that Terminal instance - see screenshot from debugger below showing the Terminal instance contains `shellArgs` and `shellPath` fields instead of `pty` as I described in the original question

![Screenshot_2024-11-13_12-17-38](https://github.com/user-attachments/assets/4ba684df-c041-4684-b685-06fffaea2c2c)


I can see the string `Hello` I write in my CustomExecution appear inside the terminal called `my test` created once I run the task so I believe I create and control the terminal correctly and the instance I'm looking at inside the listener shown in the screenshot is the `Terminal` instance I care about, but you can see there is no way to "know that" in code just from the data

**Reply by meganrogge** [reply, marked as answer]
Thanks for the info. I see what you mean. I don't think there's a way to know that at the moment.

**Comment by TheCrether** [comment]
Hi, is there any info, if this is planned to be implemented in the future?

**Reply by meganrogge** [reply]
related https://github.com/microsoft/vscode/issues/234440
